# Segmenting and Clustering Neighborhoods in Toronto

## IBM Data Science Specialztion - Coursera

## Malege Tshepo Mhozya

In [56]:
!conda install -c conda-forge beautifulsoup4 --yes

!conda install -c conda-forge geopy --yes

!conda install -c conda-forge folium=0.5.0 --yes

print('Libraries installed!')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Libraries installed!


In [57]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests
from pandas.io.json import json_normalize
import json

import requests

from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported!')

Libraries imported!


# Part 1: Web scraping for Toronto neighborhood and build a clean dataframe

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and 
    to transform the data into a pandas dataframe.

* The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
* Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
* More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.
* If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
* Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
* In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

**Note:** There are different website scraping libraries and packages in Python. One of the most common packages is BeautifulSoup. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

In [58]:
# open Wiki page with Beautiful Soup
data = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(data, 'html.parser')

**Scrape the HTML**

The data we want is in a table, with 3 columns **PostalCode, Borough and Neighborhood.**

The table contains a list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto in the province of Ontario.
Only the first three characters are listed, corresponding to the Forward Sortation Area.

In [59]:
postalCodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalCodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text.rstrip('\n')) # remove the new line char from neighborhood cell

**Put the scraped data into dataframe**

After scraping the html, we get 3 lists containing the PostalCodes, Boroughs and Neighborhoods in order as in Wiki page.
We will put those 3 lists into a dataframe with corresponding columns name.

In [60]:
toronto_neighorhood = [('PostalCode', postalCodeList),
                      ('Borough', boroughList),
                      ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_neighorhood))
toronto_df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [61]:
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_dropna.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

**Group neighborhoods by postal and borough**

There are some neighborhoods that belongs to the same postal code and borough (Ex: M5A Downtown Toronto has 2 neighborhoods Harbourfront and Regent Park).
We will concat them into same row, seperated by a colon.

In [62]:
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

**Deal with Not assigned Neighborhood**

For M7A Queen's Park, there is no neighborhood assigned.
We will replace the 'Not assigned' with the value of the corresponding Borough

In [63]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

In [64]:
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)

# Part 2: Getting coordinates and add to the Toronto DataFrame

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates.

geolocator = Nominatim(user_agent="tl-toronto-neigh")

postalList = toronto_df_cleaned['PostalCode'].values latList = [] longList = []

for post in postalList: location = None while(location is None): location = geolocator.geocode('{}, Toronto, Ontario'.format(post)) if(location != None): lat = location.latitude long = location.longitude print(post, lat, long) latList.append(lat) longList.append(long)

toronto_coors = [('Postal Code', postalList), ('Latitude', latList), ('Longitude', longList)] coors = pd.DataFrame.from_items(toronto_coors)

**Note:** Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

The csv file format has:
* 3 columns: Postal Code, Latitude and Longitude
* 103 rows: corresponding to 103 postal codes in our toronto dataframe

In [65]:
!wget -q -O "toronto_coordinates.csv" http://cocl.us/Geospatial_data
print('Coordinates downloaded!')
coors = pd.read_csv('toronto_coordinates.csv')

Coordinates downloaded!


In [66]:
print(coors.shape)
coors.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We will merge the two dataframes.
To get the correct result independant on the order of data, we need to set indexes of two dataframes to its Postal Code columns

In [67]:
toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# Part 3: Explore and cluster the neighborhoods in Toronto

In [68]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


### Using Folium to create a Map of Toronto with Boroughs markers on top

In [69]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Reduce the number of Boroughs to explore To reduce the numbers of calls to FourSquare API, we will only explore boroughs that have Toronto in their names.

In [70]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(38, 5)


PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [71]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

### Using FourSquare API to explore the Boroughs

In [72]:
CLIENT_ID = 'CBNEPAALS2RQAIZIYUJQISHLT1OPM4BJT1HL305AQY5VRDF0' # your Foursquare ID
CLIENT_SECRET = 'VEUT5ZTTMGWQZOSHXCJXR1N3FOA3UZEKQCOJFAXMI2GUBZAE' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CBNEPAALS2RQAIZIYUJQISHLT1OPM4BJT1HL305AQY5VRDF0
CLIENT_SECRET:VEUT5ZTTMGWQZOSHXCJXR1N3FOA3UZEKQCOJFAXMI2GUBZAE


In [73]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [74]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1724, 9)


PostalCode       Borough                 Neighborhood  BoroughLatitude  \
0        M4E  East Toronto                  The Beaches        43.676357   
1        M4E  East Toronto                  The Beaches        43.676357   
2        M4E  East Toronto                  The Beaches        43.676357   
3        M4E  East Toronto                  The Beaches        43.676357   
4        M4K  East Toronto  The Danforth West,Riverdale        43.679557   

   BoroughLongitude                           VenueName  VenueLatitude  \
0        -79.293031                   Glen Manor Ravine      43.676821   
1        -79.293031  The Big Carrot Natural Food Market      43.678879   
2        -79.293031                 Grover Pub and Grub      43.679181   
3        -79.293031                       Upper Beaches      43.680563   
4        -79.352188                            Pantheon      43.677621   

   VenueLongitude      VenueCategory  
0      -79.293942              Trail  
1      -79.297734  Health Food Store  
2      -79.297215                Pub  
3      -79.292869       Neighborhood  
4      -79.351434   Greek Restaurant

In [75]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                        
M4E         East Toronto      The Beaches                                                                                               4
M4K         East Toronto      The Danforth West,Riverdale                                                                              42
M4L         East Toronto      The Beaches West,India Bazaar                                                                            22
M4M         East Toronto      Studio District                                                                                          41
M4N         Central Toronto   Lawrence Park                                                                                             3
M4P         Central Toronto   Davisville North                                                                                          8
M4R         Central Toronto   North Toron

In [76]:
len(venues_df['VenueCategory'].unique())

240

#### Analyze venues in each area

In [77]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1724, 243)


PostalCode       Borough                Neighborhoods  Afghan Restaurant  \
0        M4E  East Toronto                  The Beaches                  0   
1        M4E  East Toronto                  The Beaches                  0   
2        M4E  East Toronto                  The Beaches                  0   
3        M4E  East Toronto                  The Beaches                  0   
4        M4K  East Toronto  The Danforth West,Riverdale                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0

#### Get the frequency of occurance of each category in an area

In [78]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()


(38, 243)


PostalCode          Borough                  Neighborhoods  \
0        M4E     East Toronto                    The Beaches   
1        M4K     East Toronto    The Danforth West,Riverdale   
2        M4L     East Toronto  The Beaches West,India Bazaar   
3        M4M     East Toronto                Studio District   
4        M4N  Central Toronto                  Lawrence Park   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0              0.00000   
1             0.0              0.0               0.0              0.02381   
2             0.0              0.0               0.0              0.00000   
3             0.0              0.0               0.0              0.04878   
4             0.0              0.0               0.0              0.00000   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0              0.00000   
1           0.0       0.0          0.0         0.0              0.00000   
2           0.0       0.0          0.0         0.0              0.00000   
3           0.0       0.0          0.0         0.0              0.02439   
4           0.0       0.0          0.0         0.0              0.00000   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0               0.0                 0.0            0.0        0.0         0.0   
1               0.0                 0.0            0.0        0.0         0.0   
2               0.0                 0.0            0.0        0.0         0.0   
3               0.0                 0.0            0.0        0.0         0.0   
4               0.0                 0.0            0.0        0.0         0.0   

   Bagel Shop   Bakery     Bank      Bar  Baseball Stadium  \
0         0.0  0.00000  0.00000  0.00000               0.0   
1         0.0  0.02381  0.00000  0.00000               0.0   
2         0.0  0.00000  0.00000  0.00000               0.0   
3         0.0  0.04878  0.02439  0.02439               0.0   
4         0.0  0.00000  0.00000  0.00000               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0         0.0   
1                 0.0    0.0              0.0       0.0         0.0   
2                 0.0    0.0              0.0       0.0         0.0   
3                 0.0    0.0              0.0       0.0         0.0   
4                 0.0    0.0              0.0       0.0         0.0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0     0.0            0.0    0.00000       0.0   
1                 0.0     0.0            0.0    0.02381       0.0   
2                 0.0     0.0            0.0    0.00000       0.0   
3                 0.0     0.0            0.0    0.02439       0.0   
4                 0.0     0.0            0.0    0.00000       0.0   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  Building  \
0                   0.0             0.0  0.000000          0.00000       0.0   
1                   0.0             0.0  0.023810          0.02381       0.0   
2                   0.0             0.0  0.045455          0.00000       0.0   
3                   0.0             0.0  0.024390          0.00000       0.0   
4                   0.0             0.0  0.000000          0.00000       0.0   

   Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
0      0.000000       0.000000  0.000000      0.0  0.0

#### Get 10 most occurance venue types in each area

In [79]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
35        M6R      West Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
36        M6S      West Toronto   
30        M6G  Downtown Toronto   
24        M5R   Central Toronto   
18        M5H  Downtown Toronto   
21        M5L  Downtown Toronto   
15        M5C  Downtown Toronto   
20        M5K  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
14        M5B  Downtown Toronto   
16        M5E  Downtown Toronto   
11        M4X  Downtown Toronto   
19        M5J  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
9         M4V   Central Toronto   
7         M4S   Central Toronto   
1         M4K      East Toronto   
0         M4E      East Toronto   
22        M5N   Central Toronto   
34        M6P      West Toronto   
10        M4W  Downtown Toronto   
2         M4L      East Toronto   
4         M4N   Central Toronto   
31        M6H      West Toronto   
8         M4T   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
23        M5P   Central Toronto   
37        M7Y      East Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   
32                            Little Portugal,Trinity                   Bar   
35                              Parkdale,Roncesvalles        Breakfast Spot   
25                      Harbord,University of Toronto                  Café   
26            Chinatown,Grange Park,Kensington Market                  Café   
33         Brockton,Exhibition Place,Parkdale Village                  Café   
3                                     Studio District                  Café   
36                                  Runnymede,Swansea                  Café   
30                                           Christie                  Café   
24                  The Annex,North Midtown,Yorkville                  Café   
18                             Adelaide,King,Richmond           Coffee Shop   
21                      Commerce Court,Victoria Hotel           Coffee Shop   
15                                     St. James Town           Coffee Shop   
20            Design Exchange,Toronto Dominion Centre           Coffee Shop   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
29              First Canadian Place,Underground city           Coffee Shop   
14                            Ryerson,Garden District           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
11                         Cabbagetown,St. James Town           Coffee Shop   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                           Harbourfront,Regent Park           Coffee Shop   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...           Coffee Shop   
7                                          Davisville          Dessert Shop   
1                         The Danforth West,Riverdale      Greek Restaurant   
0                                         The Beaches     Health Food Store   
22                                           Roselawn          Home Service   
34                       High Park,The Junction South    Mexican Restaurant   
10                                           Rosedale                  Park   
2                       The Beaches West,India Bazaar                  Park   
4                                       Lawrence Park                  Park   
31                        Dove

**Clustering areas**

Use KMeans algorigthm, try to cluster the toronto central areas into 4 clusters

In [80]:
kclusters = 4

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
35        M6R      West Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
36        M6S      West Toronto   
30        M6G  Downtown Toronto   
24        M5R   Central Toronto   
18        M5H  Downtown Toronto   
21        M5L  Downtown Toronto   
15        M5C  Downtown Toronto   
20        M5K  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
14        M5B  Downtown Toronto   
16        M5E  Downtown Toronto   
11        M4X  Downtown Toronto   
19        M5J  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
9         M4V   Central Toronto   
7         M4S   Central Toronto   
1         M4K      East Toronto   
0         M4E      East Toronto   
34        M6P      West Toronto   
2         M4L      East Toronto   
31        M6H      West Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
37        M7Y      East Toronto   
10        M4W  Downtown Toronto   
4         M4N   Central Toronto   
23        M5P   Central Toronto   
8         M4T   Central Toronto   
22        M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
32                            Little Portugal,Trinity  43.647927 -79.419750   
35                              Parkdale,Roncesvalles  43.648960 -79.456325   
25                      Harbord,University of Toronto  43.662696 -79.400049   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
33         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191   
3                                     Studio District  43.659526 -79.340923   
36                                  Runnymede,Swansea  43.651571 -79.484450   
30                                           Christie  43.669542 -79.422564   
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
15                                     St. James Town  43.651494 -79.375418   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
29              First Canadian Place,Underground city  43.648429 -79.382280   
14                            Ryerson,Garden District  43.657162 -79.378937   
16                                        Berczy Park  43.644771 -79.373306   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
17                                 Central Bay Street  43.657952 -79.387383   
12                               Church and Wellesley  43.665860 -79.383160   
13                           Harbourfront,Regent Park  43.654260 -79.360636   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049   
7                                          Davisville  43.704324 -79.388790   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
0                                         The Beaches  43.676357 -79.293031   
34                       High Park,The Junction South  43.661608 -79.464763   
2                       The Beaches West,India Bazaar  43.668999 -79.315572   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
5                                    Davisville North  43.712751 -79.390197   
6                                  North Toronto West  43.715383 -79.405678   
37  Business Reply Mail Proces

#### Show those clusters onto a map

In [81]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

**Upon observing the result, we can name the clusters as follow:**

* Cluster 0 (Red): Living area (with mostly park, trail, school; and some small businesses)
* Cluster 1 (Yellow): Roselawn - Central Toronto (nothing here except a garden)
* Cluster 2 (Purple): Business area (with lots of business venues)
* Cluster 3 (Blue) : Oldest neighboorhood with lot of nature park

Thank for following my project.